File: Graphs/wordLadder.py
from collections import deque

class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        
        def construct_dict(word_list):
            dict = {}
            for word in word_list:
                for i in range(len(word)):
                    pattern = word[:i] + "_" + word[i+1:]
                    dict[pattern] = dict.get(pattern, []) + [word]
            return dict
            
        def bfs_words(begin, end, dict_words):
            queue, visited = deque([(begin, 1)]), set()
            while queue:
                word, depth = queue.popleft()
                if word not in visited:
                    if word == end:
                        return depth
                    visited.add(word)
                    for i in range(len(word)):
                        pattern = word[:i] + "_" + word[i+1:]
                        neigh_words = dict_words.get(pattern, [])
                        for neighbor in neigh_words:
                            if neighbor not in visited:
                                queue.append((neighbor, depth + 1))
            return 0
        
        dict = construct_dict(set(wordList))
        return bfs_words(beginWord, endWord, dict)

File: Graphs/wordLadder.js
const ladderLength = (beginWord, endWord, wordList) => {
    let transformations = 1
    let queue = [beginWord]
    let dictionary = new Set(wordList)
    let visited = new Set()
    
    while (queue.length) {
        const neighbors = []
        for (let currentWord of queue) {
            if (currentWord === endWord) {
                return transformations
            }
            let currentWordArr = currentWord.split('')
            for (let i = 0; i < currentWordArr.length; i++) {
                for (let j = 0; j < 26; j++) {
                    currentWordArr[i] = String.fromCharCode(97 + j)
                    const newCurrent = currentWordArr.join('')
                    if (dictionary.has(newCurrent) && !visited.has(newCurrent)) {
                        neighbors.push(newCurrent)
                        visited.add(newCurrent)
                    }
                    currentWordArr[i] = currentWord[i]
                }
            }
        }
        queue = neighbors
        transformations++
    }
    return 0
};